src/modules/ekf2/EKF/height_control.cpp

```diff
@@ -0,0 +1,235 @@
+
+	Likelihood inertial_nav_falling_likelihood = estimateInertialNavFallingLikelihood();
+
+	// Check for more than 50% clipping affected IMU samples within the past 1 second
+	const uint16_t clip_count_limit = 1.f / _dt_ekf_avg;
+	const bool is_clipping = _imu_sample_delayed.delta_vel_clipping[0] ||
+				 _imu_sample_delayed.delta_vel_clipping[1] ||
+				 _imu_sample_delayed.delta_vel_clipping[2];
+
+	if (is_clipping && _clip_counter < clip_count_limit) {
+		_clip_counter++;
+
+	} else if (_clip_counter > 0) {
+		_clip_counter--;
+	}
+
+	_fault_status.flags.bad_acc_clipping = _clip_counter > clip_count_limit / 2;
+
+	const bool is_clipping_frequently = _clip_counter > 0;
+
+	// Do not require evidence of clipping if the likelihood of having the INS falling is high
+	const bool bad_vert_accel = (is_clipping_frequently && (inertial_nav_falling_likelihood == Likelihood::MEDIUM))
+				    || (inertial_nav_falling_likelihood == Likelihood::HIGH);
+
+	if (bad_vert_accel) {
+		_time_bad_vert_accel = _time_last_imu;
+
+	} else {
+		_time_good_vert_accel = _time_last_imu;
+	}
+
+	// declare a bad vertical acceleration measurement and make the declaration persist
+	// for a minimum of BADACC_PROBATION seconds
+	if (_fault_status.flags.bad_acc_vertical) {
+		_fault_status.flags.bad_acc_vertical = isRecent(_time_bad_vert_accel, BADACC_PROBATION);
+
+	} else {
+		_fault_status.flags.bad_acc_vertical = bad_vert_accel;
+	}
+}
+
+Likelihood Ekf::estimateInertialNavFallingLikelihood() const
+{
+	bool likelihood_high = false;
+	bool likelihood_medium = false;
+
+	enum class ReferenceType { PRESSURE, GNSS, GROUND };
+
+	struct {
+		ReferenceType ref_type;
+		float innov;
+		float innov_var;
+		bool failed_min;
+		bool failed_lim;
+	} checks[6] {};
+
+	if (_control_status.flags.baro_hgt) {
+		checks[0] = {ReferenceType::PRESSURE, _aid_src_baro_hgt.innovation, _aid_src_baro_hgt.innovation_variance};
+	}
+
+	if (_control_status.flags.gps_hgt) {
+		checks[1] = {ReferenceType::GNSS, _aid_src_gnss_pos.innovation[2], _aid_src_gnss_pos.innovation_variance[2]};
+	}
+
+	if (_control_status.flags.gps) {
+		checks[2] = {ReferenceType::GNSS, _aid_src_gnss_vel.innovation[2], _aid_src_gnss_vel.innovation_variance[2]};
+	}
+
+	if (_control_status.flags.rng_hgt) {
+		checks[3] = {ReferenceType::GROUND, _aid_src_rng_hgt.innovation, _aid_src_rng_hgt.innovation_variance};
+	}
+
+	if (_control_status.flags.ev_hgt) {
+		checks[4] = {ReferenceType::GROUND, _ev_pos_innov(2), _ev_pos_innov_var(2)};
+	}
+
+	if (_control_status.flags.ev_vel) {
+		checks[5] = {ReferenceType::GROUND, _ev_vel_innov(2), _ev_vel_innov_var(2)};
+	}
+
+	// Compute the check based on innovation ratio for all the sources
+	for (unsigned i = 0; i < 6; i++) {
+		if (checks[i].innov_var < FLT_EPSILON) {
+			continue;
+		}
+
+		const float innov_ratio = checks[i].innov / sqrtf(checks[i].innov_var);
+		checks[i].failed_min = innov_ratio > _params.vert_innov_test_min;
```

Comment:
For convenience we could add innovation ratio to estimator_aid_source.

Reply:
I didn't want to change too much but since this is a fixed threshold, we could use the "signed normalized innovation squared" (i.e.: sign(innov) * test_ratio * gate) to avoid the `sqrtf` or even the "signed test ratio"